revocation/crl/fetcher/fetcher.go

```diff
@@ -0,0 +1,155 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package fetcher provides Fetcher interface and its implementation to fetch
+// CRL from the given URL
+package fetcher
+
+import (
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/notaryproject/notation-core-go/revocation/crl/cache"
+)
+
+// maxCRLSize is the maximum size of CRL in bytes
+const maxCRLSize = 64 * 1024 * 1024 // 64 MiB
+
+// Fetcher is an interface that specifies methods used for fetching CRL
+// from the given URL
+type Fetcher interface {
+	Fetch(ctx context.Context, crlURL string) (bundle *cache.Bundle, fromCache bool, err error)
+}
+
+type cachedFetcher struct {
+	httpClient  *http.Client
+	cacheClient cache.Cache
+}
+
+// NewCachedFetcher creates a new Fetcher with the given HTTP client and cache client
+//   - if httpClient is nil, http.DefaultClient will be used
+//   - if cacheClient is nil, no cache will be used
+func NewCachedFetcher(httpClient *http.Client, cacheClient cache.Cache) (Fetcher, error) {
```

Comment:
Shall we make Fetcher a singleton to avoid concurrency issues?

Reply:
Could you please explain the concurrency issues? Which parts may trigger a concurrency issue?

Reply:
As discussed, concurrency is resolved.